### DISCUSSION TITLE
Конвертация кода от JavaScript к TypeScript

### DISCUSSION THREAD

**Discussion by gregor-shkashekwa** [VKCOM/VKUI#7282] [Q&A] [answered]
Здравствуйте, я столкнулся с одной проблемой при конвертации кода от JavaScript к TypeScript в своем мини-приложении ВК. В частности, это связано с тем, что TypeScript не распознает компоненты VKUI.

К примеру, я приступил к конвертированию компонента index.js в index.ts. При отладке споткнулся об ошибки:
![Скриншот 02-08-2024 005657](https://github.com/user-attachments/assets/de18540f-3bf5-4705-8da5-fd359cf5b2c2)
ERROR
Module not found: Error: Can't resolve 'C:\Users\User\Desktop\resume_constructor\src\index.js' in 'C:\Users\User\Desktop\resume_constructor'
ERROR in src/index.ts:10:34
TS2345: Argument of type 'HTMLElement | null' is not assignable to parameter of type 'Container'.
  Type 'null' is not assignable to type 'Container'.
     8 | bridge.send("VKWebAppInit");
     9 |
  > 10 | const root = ReactDOM.createRoot(document.getElementById('root'));
       |                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    11 | root.render(
    12 |     <ConfigProvider>
    13 |         <AdaptivityProvider>
ERROR in src/index.ts:12:5
TS2365: Operator '<' cannot be applied to types 'boolean' and 'RegExp'.
    10 | const root = ReactDOM.createRoot(document.getElementById('root'));
    11 | root.render(
  > 12 |     <ConfigProvider>
       |     ^^^^^^^^^^^^^^^^
  > 13 |         <AdaptivityProvider>
       | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  > 14 |             <App />
       | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  > 15 |         </AdaptivityProvider>
       | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  > 16 |     </ConfigProvider>
       | ^^^^^^^^^^^^^^^^^^^^^^
    17 | );
ERROR in src/index.ts:12:6
TS2749: 'ConfigProvider' refers to a value, but is being used as a type here. Did you mean 'typeof ConfigProvider'?
    10 | const root = ReactDOM.createRoot(document.getElementById('root'));
    11 | root.render(
  > 12 |     <ConfigProvider>
       |      ^^^^^^^^^^^^^^
    13 |         <AdaptivityProvider>
    14 |             <App />
    15 |         </AdaptivityProvider>
ERROR in src/index.ts:13:10
TS2749: 'AdaptivityProvider' refers to a value, but is being used as a type here. Did you mean 'typeof AdaptivityProvider'?
    11 | root.render(
    12 |     <ConfigProvider>
  > 13 |         <AdaptivityProvider>
       |          ^^^^^^^^^^^^^^^^^^
    14 |             <App />
    15 |         </AdaptivityProvider>
    16 |     </ConfigProvider>
ERROR in src/index.ts:14:14
TS2749: 'App' refers to a value, but is being used as a type here. Did you mean 'typeof App'?
    12 |     <ConfigProvider>
    13 |         <AdaptivityProvider>
  > 14 |             <App />
       |              ^^^
    15 |         </AdaptivityProvider>
    16 |     </ConfigProvider>
    17 | );
ERROR in src/index.ts:14:18
TS1005: '>' expected.
    12 |     <ConfigProvider>
    13 |         <AdaptivityProvider>
  > 14 |             <App />
       |                  ^
    15 |         </AdaptivityProvider>
    16 |     </ConfigProvider>
    17 | );
ERROR in src/index.ts:15:10
TS1161: Unterminated regular expression literal.
    13 |         <AdaptivityProvider>
    14 |             <App />
  > 15 |         </AdaptivityProvider>
       |          ^^^^^^^^^^^^^^^^^^^^
    16 |     </ConfigProvider>
    17 | );
ERROR in src/index.ts:16:6
TS1161: Unterminated regular expression literal.
    14 |             <App />
    15 |         </AdaptivityProvider>
  > 16 |     </ConfigProvider>
       |      ^^^^^^^^^^^^^^^^
    17 | );
    
    Вопрос: что нужно настроить в проекте, чтобы TypeScript распознавал компоненты VKUI?

**Comment by SevereCloud** [comment, marked as answer]
**TLDR:** переименуй файл из index.ts в index.ts**x**

---

Привет, typescript ругается не на VKUI, а на неверный синтаксис. Typescript ожидает ts код, но натыкается на синтаксис [jsx](https://react.dev/learn/writing-markup-with-jsx). 

> TS2365: Operator '<' cannot be applied to types 'boolean' and 'RegExp'.

Чтобы использовать синтаксис [jsx](https://www.typescriptlang.org/docs/handbook/jsx.html) в typescript требуется две вещи:

1. Файлы с расширением `.tsx`
2. Настройка опции [jsx](https://www.typescriptlang.org/tsconfig#jsx)

**Reply by gregor-shkashekwa** [reply]
Во, прекрасно заработало, спасибо большое)
